### Multi Layer Perceptron - TON IoT Dataset

### Data Preprocessing

In [92]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [93]:
df = pd.read_csv("/content/drive/MyDrive/ML Lab/LAB_ToN_IoT.csv")
df.head()

Unnamed: 0                                 Flow ID         Src IP  \
0     1909167   176.28.50.165-192.168.1.32-80-49056-6  176.28.50.165   
1     5064909   52.28.231.150-192.168.1.35-80-58548-6  52.28.231.150   
2      589938  192.168.1.193-192.168.1.33-445-33721-6  192.168.1.193   
3     1126246   52.28.231.150-192.168.1.36-80-34034-6  52.28.231.150   
4     1827801   192.168.1.190-192.168.1.36-80-41390-6  192.168.1.190   

   Src Port        Dst IP  Dst Port  Protocol               Timestamp  \
0        80  192.168.1.32     49056         6  27/04/2019 04:47:51 pm   
1        80  192.168.1.35     58548         6  27/04/2019 10:37:54 pm   
2       445  192.168.1.33     33721         6  28/04/2019 02:57:14 pm   
3        80  192.168.1.36     34034         6  27/04/2019 03:51:29 pm   
4        80  192.168.1.36     41390         6  27/04/2019 04:39:33 pm   

   Flow Duration  Tot Fwd Pkts  ...  Active Mean  Active Std  Active Max  \
0         801869             5  ...          0.0         0.0         0.0   
1            182             1  ...          0.0         0.0         0.0   
2            313             2  ...          0.0         0.0         0.0   
3            386             1  ...          0.0         0.0         0.0   
4          24199             4  ...          0.0         0.0         0.0   

   Active Min     Idle Mean  Idle Std      Idle Max      Idle Min  Label  \
0         0.0  1.556348e+15       0.0  1.556348e+15  1.556348e+15      0   
1         0.0  1.556369e+15       0.0  1.556369e+15  1.556369e+15      0   
2         0.0  1.556427e+15       0.0  1.556427e+15  1.556427e+15      0   
3         0.0  1.556344e+15       0.0  1.556344e+15  1.556344e+15      0   
4         0.0  1.556347e+15       0.0  1.556347e+15  1.556347e+15      0   

   Attack  
0  Benign  
1  Benign  
2  Benign  
3  Benign  
4  Benign  

[5 rows x 86 columns]

In [94]:
df.shape

(133794, 86)

In [95]:
df.columns

Index(['Unnamed: 0', 'Flow ID', 'Src IP', 'Src Port', 'Dst IP', 'Dst Port',
       'Protocol', 'Timestamp', 'Flow Duration', 'Tot Fwd Pkts',
       'Tot Bwd Pkts', 'TotLen Fwd Pkts', 'TotLen Bwd Pkts', 'Fwd Pkt Len Max',
       'Fwd Pkt Len Min', 'Fwd Pkt Len Mean', 'Fwd Pkt Len Std',
       'Bwd Pkt Len Max', 'Bwd Pkt Len Min', 'Bwd Pkt Len Mean',
       'Bwd Pkt Len Std', 'Flow Byts/s', 'Flow Pkts/s', 'Flow IAT Mean',
       'Flow IAT Std', 'Flow IAT Max', 'Flow IAT Min', 'Fwd IAT Tot',
       'Fwd IAT Mean', 'Fwd IAT Std', 'Fwd IAT Max', 'Fwd IAT Min',
       'Bwd IAT Tot', 'Bwd IAT Mean', 'Bwd IAT Std', 'Bwd IAT Max',
       'Bwd IAT Min', 'Fwd PSH Flags', 'Bwd PSH Flags', 'Fwd URG Flags',
       'Bwd URG Flags', 'Fwd Header Len', 'Bwd Header Len', 'Fwd Pkts/s',
       'Bwd Pkts/s', 'Pkt Len Min', 'Pkt Len Max', 'Pkt Len Mean',
       'Pkt Len Std', 'Pkt Len Var', 'FIN Flag Cnt', 'SYN Flag Cnt',
       'RST Flag Cnt', 'PSH Flag Cnt', 'ACK Flag Cnt', 'URG Flag Cnt',
       'CWE Flag

In [96]:
df['Attack'].unique()

array(['Benign', 'backdoor', 'ddos', 'dos', 'injection', 'mitm',
       'password', 'ransomware', 'scanning', 'xss'], dtype=object)

In [97]:
columns_to_remove=['Unnamed: 0','Flow ID','Src IP','Src Port','Dst IP', 'Dst Port','Timestamp']
df = df.drop(columns=columns_to_remove)
df.columns

Index(['Protocol', 'Flow Duration', 'Tot Fwd Pkts', 'Tot Bwd Pkts',
       'TotLen Fwd Pkts', 'TotLen Bwd Pkts', 'Fwd Pkt Len Max',
       'Fwd Pkt Len Min', 'Fwd Pkt Len Mean', 'Fwd Pkt Len Std',
       'Bwd Pkt Len Max', 'Bwd Pkt Len Min', 'Bwd Pkt Len Mean',
       'Bwd Pkt Len Std', 'Flow Byts/s', 'Flow Pkts/s', 'Flow IAT Mean',
       'Flow IAT Std', 'Flow IAT Max', 'Flow IAT Min', 'Fwd IAT Tot',
       'Fwd IAT Mean', 'Fwd IAT Std', 'Fwd IAT Max', 'Fwd IAT Min',
       'Bwd IAT Tot', 'Bwd IAT Mean', 'Bwd IAT Std', 'Bwd IAT Max',
       'Bwd IAT Min', 'Fwd PSH Flags', 'Bwd PSH Flags', 'Fwd URG Flags',
       'Bwd URG Flags', 'Fwd Header Len', 'Bwd Header Len', 'Fwd Pkts/s',
       'Bwd Pkts/s', 'Pkt Len Min', 'Pkt Len Max', 'Pkt Len Mean',
       'Pkt Len Std', 'Pkt Len Var', 'FIN Flag Cnt', 'SYN Flag Cnt',
       'RST Flag Cnt', 'PSH Flag Cnt', 'ACK Flag Cnt', 'URG Flag Cnt',
       'CWE Flag Count', 'ECE Flag Cnt', 'Down/Up Ratio', 'Pkt Size Avg',
       'Fwd Seg Size Avg', 'Bwd

In [98]:
df.isnull().sum()[df.isnull().sum() > 0]

Series([], dtype: int64)

In [99]:
numeric_cols = df.select_dtypes(include=[np.number]).columns
print(np.isinf(df[numeric_cols]).sum())
df[numeric_cols] = df[numeric_cols].replace([np.inf, -np.inf], np.nan)
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].median())

Protocol           0
Flow Duration      0
Tot Fwd Pkts       0
Tot Bwd Pkts       0
TotLen Fwd Pkts    0
                  ..
Idle Mean          0
Idle Std           0
Idle Max           0
Idle Min           0
Label              0
Length: 78, dtype: int64


### Binary Classification
- Attack(1) or No Attack(0)

In [100]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

In [101]:
X = df.drop(columns=["Label","Attack"])
y = df["Label"]


In [102]:
np.isinf(X).sum()
X.replace([np.inf, -np.inf], np.nan, inplace=True)
X.fillna(X.median(), inplace=True)

In [103]:
X_sample = X.sample(10000, random_state=1010)  # 10k rows
y_sample = y.loc[X_sample.index]
scaler = StandardScaler()
X_train, X_test, y_train, y_test = train_test_split(
    X_sample, y_sample, test_size=0.2, random_state=1010, stratify=y_sample
)
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


- Activation Function: Relu

In [104]:
mlp = MLPClassifier(hidden_layer_sizes=(64,32),max_iter=200,random_state=1010,activation='relu',solver='adam')
mlp.fit(X_train,y_train)

y_pred = mlp.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)

print(f"Accuracy: {accuracy}")

Accuracy: 0.9745


- Activation Function: tanh

In [105]:
mlp = MLPClassifier(hidden_layer_sizes=(64,32),max_iter=200,random_state=1010,activation='tanh',solver='adam')
mlp.fit(X_train,y_train)

y_pred = mlp.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)

print(f"Accuracy: {accuracy}")

Accuracy: 0.9725


- Activation Function: logistic (sigmoid)

In [115]:
mlp = MLPClassifier(hidden_layer_sizes=(64,32),max_iter=500,random_state=1010,activation='logistic',solver='adam')
mlp.fit(X_train,y_train)

y_pred = mlp.predict(X_test)
accuracy = accuracy_score(y_test,y_pred)

print(f"Accuracy: {accuracy}")

Accuracy: 0.968


### Multiclass Classification ( 0 - 8 )

In [119]:
from sklearn.preprocessing import LabelEncoder

# X = features
X = df.drop(columns=['Label', 'Attack'])
# y = Attack column
y = df['Attack']

# Encode categorical labels to numbers
le = LabelEncoder()
y_encoded = le.fit_transform(y)

In [120]:
import numpy as np

X.replace([np.inf, -np.inf], np.nan, inplace=True)
X.fillna(X.median(), inplace=True)

In [121]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y_encoded, test_size=0.2, random_state=42, stratify=y_encoded
)

In [123]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [122]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report

mlp = MLPClassifier(
    hidden_layer_sizes=(128, 64),
    activation='relu',
    solver='adam',
    max_iter=300,
    random_state=42
)

# Train
mlp.fit(X_train, y_train)

# Predict
y_pred = mlp.predict(X_test)

# Evaluate
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred, target_names=le.classes_))

Accuracy: 0.46074218020105384
              precision    recall  f1-score   support

      Benign       0.82      0.15      0.25     12576
    backdoor       0.00      0.00      0.00       129
        ddos       0.00      0.00      0.00         1
         dos       0.00      0.00      0.00         1
   injection       0.00      0.00      0.00      1416
        mitm       0.00      0.00      0.00         2
    password       0.00      0.00      0.00      1695
  ransomware       0.00      0.00      0.00        23
    scanning       0.00      0.00      0.00       185
         xss       0.43      0.98      0.59     10731

    accuracy                           0.46     26759
   macro avg       0.13      0.11      0.08     26759
weighted avg       0.56      0.46      0.36     26759



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


# Conclusion

1. **Binary Classification (Label: 0 or 1)**

   * MLP with two hidden layers `(64,32)` performs **very well**:

     * ReLU: 97.45%
     * Tanh: 97.25%
     * Sigmoid: 96.8%
   * All three activations give high accuracy, showing that the model can reliably distinguish **normal vs attack traffic**.
   * ReLU slightly outperforms the others, likely due to better gradient flow and faster convergence.

2. **Multi-class Classification (Attack types)**

   * MLP struggles with **specific attack types**, achieving only \~46% accuracy.
   * Metrics reveal:

     * Majority classes (e.g., `Benign`, `xss`) dominate predictions.
     * Minority classes (e.g., `ddos`, `dos`, `mitm`) are almost never predicted → **very low recall and F1-score**.
   * Challenges arise from **class imbalance**, very few samples for some attacks, and limited feature separation for multiple attack types.

## Justification

* MLP works well for **binary detection** because the classes are clearly separable.
* For **multi-class detection**, the model struggles without additional strategies like:

  * Class balancing (oversampling minority classes)
  * Feature engineering to better separate attack types
  * Larger/deeper networks or alternative models for complex patterns.

Overall, the pre-defined MLP demonstrates **strong capability for binary attack detection** but highlights **limitations for multi-class classification** on imbalanced IoT/CICIDS datasets.
